### DISCUSSION TITLE
How does type inference on return values work / should it work here?

### DISCUSSION THREAD

**Discussion by mds1** [wevm/viem#468] [Question] [answered]
Hey! Finally starting to use viem a bit more and it's really nice so far. I'm working on some multicall examples using viem, and had some questions around type inference. The file I'll be referencing is [here](https://github.com/mds1/multicall/blob/449973755bdd9ff903c254faefeaa770a7bc38c3/examples/typescript/viem.ts).

- For the multicall on [line 35](https://github.com/mds1/multicall/blob/449973755bdd9ff903c254faefeaa770a7bc38c3/examples/typescript/viem.ts#L35), the return values have a type of `unknown`, but I'd expect them all be a `0x${string}` 
- For the multicall on [line 46](https://github.com/mds1/multicall/blob/449973755bdd9ff903c254faefeaa770a7bc38c3/examples/typescript/viem.ts#L46), the type inference is a little better. You can see the details in the comment above, but it seems the expected type is the union of all types returned, as opposed to a specific type based on call index
- On [line 68](https://github.com/mds1/multicall/blob/449973755bdd9ff903c254faefeaa770a7bc38c3/examples/typescript/viem.ts#L68), the type inference works as expected for that single call (not a multicall).

I haven't used viem much, so definitely a chance I'm configuring something wrong here, but I'm basically wondering;
1. Is the above behavior all expected?
2. If not, is it because I'm configuring something wrong, or due to a bug?

The difference in type inference between the two multicall examples makes me think the answer is "I have the first multicall configured wrong, and the second multicall is a limitation of how the type inference works for multicalls", but I haven't been able to get multicall type inference working like bullet 3 yet

**Comment by jxom** [comment, marked as answer]
Raised a PR: https://github.com/mds1/multicall/pull/82

**Reply by mds1** [reply]
Amazing, thanks!
